src/replication.c

```diff
@@ -2359,6 +2381,14 @@ void replicationHandleMasterDisconnection(void) {
     /* We lost connection with our master, don't disconnect slaves yet,
      * maybe we'll be able to PSYNC with our master later. We'll disconnect
      * the slaves only if we'll have to do a full resync with our master. */
+
+    /* Try to re-connect immediately rather than wait for replicationCron
+     * waiting 1 second may risk backlog being recycled. */
+    serverLog(LL_NOTICE,"Reconnecting to MASTER %s:%d",
+        server.masterhost, server.masterport);
+    if (connectWithMaster() == C_OK) {
+        serverLog(LL_NOTICE,"MASTER <-> REPLICA sync started");
+    }
```

Comment:
The anti-code duplication in me thinks we should just move the log into connectWithMaster(), and have it return nothing. Although this part of the code is kind of all over the place.